File: core/shared/src/test/scala/diplomacy/variant/standard/map/MapSpec.scala
package diplomacy.variant.standard.map

import diplomacy.UnitSpec
import diplomacy.variant.standard.rule._
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.rule.Keywords._

class MapSpec extends UnitSpec {
  "A map instance" should {
    "define provinces (1)" in {
      Pic.province.isSupplyCenter should be(false)
      Pic.province.homeOf should be(Option(France))

      Bre.province.isSupplyCenter should be(true)
      Bre.province.homeOf should be(Option(France))

      Bur.province.isSupplyCenter should be(false)
      Bur.province.homeOf should be(Option(France))

      Gas.province.isSupplyCenter should be(false)
      Gas.province.homeOf should be(Option(France))

      map.movableLocationsOf(Par, Army) should be(Set(Pic, Bre, Bur, Gas))
    }
    "define provinces (2)" in {
      Iri.province.isSupplyCenter should be(false)
      Iri.province.homeOf should be(None)

      Wal.province.homeOf should be(Option(England))

      map.movableLocationsOf(Eng, Fleet) should be(
        Set(Iri, Wal, Lon, Nth, Bel, Pic, Bre, Mid)
      )
    }
    "define provinces (3)" in {
      map.movableLocationsOf(Rom, Fleet) should be(Set(Tus, Nap, Tyn))
    }
  }
}

File: test/variant/standard/map/map-spec.ts
import * as chai from "chai"
import { MilitaryBranch } from "./../../../../src/variant/standard/rule/module"
import { locations as $ } from "./../../../../src/variant/standard/map/location"
import { Power } from "./../../../../src/variant/standard/map/power"
import { map } from "./../../../../src/variant/standard/map/map"

const { Army, Fleet } = MilitaryBranch
const should = chai.should()

describe("Map", () => {
  it("defines a map (1).", () => {
    $.Pic.province.isSupplyCenter.should.equal(false)
    should.equal($.Pic.province.homeOf, Power.France)
    $.Bre.province.isSupplyCenter.should.equal(true)
    should.equal($.Bre.province.homeOf, Power.France)
    $.Bur.province.isSupplyCenter.should.equal(false)
    should.equal($.Bur.province.homeOf, Power.France)
    $.Gas.province.isSupplyCenter.should.equal(false)
    should.equal($.Gas.province.homeOf, Power.France)

    const provinces = map.movableProvincesOf($.Par.province, Army)

    provinces.size.should.equal(4)
    provinces.should.deep.equal(
      new Set([$.Pic.province, $.Bre.province, $.Bur.province, $.Gas.province])
    )
  })
  it("defines a map (2).", () => {
    ($.Iri.province.homeOf == null).should.equal(true)
    $.Iri.province.isSupplyCenter.should.equal(false)
    should.equal($.Wal.province.homeOf, Power.England)

    const provinces = map.movableProvincesOf($.Eng.province, Fleet)

    provinces.should.deep.equal(
      new Set([
        $.Iri.province, $.Wal.province, $.Lon.province, $.Nth.province, $.Bel.province,
        $.Pic.province, $.Bre.province, $.Mid.province
      ])
    )
  })
  it("defines a map (3).", () => {
    const provinces = map.movableProvincesOf($.Rom.province, Fleet)

    provinces.size.should.equal(3)
    provinces.should.deep.equal(new Set([
      $.Tus.province, $.Nap.province, $.Tyn.province
    ]))
  })
})
